core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -824,11 +825,15 @@ private[spark] class MapOutputTrackerWorker(conf: SparkConf) extends MapOutputTr
       endPartition: Int): Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])] = {
     logDebug(s"Fetching outputs for shuffle $shuffleId, mappers $startMapIndex-$endMapIndex" +
       s"partitions $startPartition-$endPartition")
-    val statuses = getStatuses(shuffleId, conf)
     try {
+      val statuses = getStatuses(shuffleId, conf)
       MapOutputTracker.convertMapStatuses(
         shuffleId, startPartition, endPartition, statuses, startMapIndex, endMapIndex)
     } catch {
+      case e: IOException if
+        Throwables.getCausalChain(e).asScala.exists(_.isInstanceOf[BlockNotFoundException]) =>
+        mapStatuses.clear()
```

Comment:
It looks more consistent to me to throw `MetadataFetchFailedException` in `getStatuses`. Then we can reuse the handling of `MetadataFetchFailedException` below.

Reply:
@cloud-fan do you mean put these codes in `getStatuses` ?  Yes, I think it's more clear. I'll update.